src/mbgl/style/source_impl.cpp

```diff
@@ -111,8 +126,10 @@ void Source::Impl::updateTiles(const UpdateParameters& parameters) {
     std::set<OverscaledTileID> retain;
 
     auto retainTileFn = [&retain](Tile& tile, Resource::Necessity necessity) -> void {
-        retain.emplace(tile.id);
-        tile.setNecessity(necessity);
+        if (retain.find(tile.id) == retain.end()) {
+            retain.emplace(tile.id);
+            tile.setNecessity(necessity);
+        }
```

Comment:
```cpp
if (retain.emplace(tile.id).second) {
    tile.setNecessity(necessity);
}
```

 `.second` is `true` when something was inserted, `false` otherwise.